OpenRA.Mods.Common/Traits/AutoCrusher.cs

```diff
@@ -0,0 +1,82 @@
+﻿#region Copyright & License Information
+/*
+ * Copyright 2007-2022 The OpenRA Developers (see AUTHORS)
+ * This file is part of OpenRA, which is free software. It is made
+ * available to you under the terms of the GNU General Public License
+ * as published by the Free Software Foundation, either version 3 of
+ * the License, or (at your option) any later version. For more
+ * information, see COPYING.
+ */
+#endregion
+
+using System.Linq;
+using OpenRA.Primitives;
+using OpenRA.Traits;
+
+namespace OpenRA.Mods.Common.Traits
+{
+	class AutoCrusherInfo : PausableConditionalTraitInfo
+	{
+		[Desc("Maximum scan range for AutoCrusher.")]
+		public readonly WDist ScanRadius = WDist.FromCells(5);
+
+		[Desc("Ticks to wait until next AutoCrusher: attempt.")]
+		public readonly int MinimumScanTimeInterval = 10;
+
+		[Desc("Ticks to wait until next AutoCrusher: attempt.")]
+		public readonly int MaximumScanTimeInterval = 15;
+
+		[Desc("Relationships between actor's and target's owner needed for AutoCrusher.")]
+		public readonly PlayerRelationship TargetRelationships = PlayerRelationship.Ally | PlayerRelationship.Neutral | PlayerRelationship.Enemy;
+
+		public override object Create(ActorInitializer init) { return new AutoCrusher(this); }
+	}
+
+	class AutoCrusher : PausableConditionalTrait<AutoCrusherInfo>, INotifyIdle
+	{
+		BitSet<CrushClass> crushes;
+		bool isAircraft;
+		int nextScanTime;
+		IResolveOrder move;
+
+		public AutoCrusher(AutoCrusherInfo info)
+			: base(info) { }
+
+		protected override void Created(Actor self)
+		{
+			if (self.Info.HasTraitInfo<MobileInfo>())
+				crushes = self.Info.TraitInfos<MobileInfo>().First().LocomotorInfo.Crushes;
```

Comment:
You can save a trait lookup this way:

```suggestion
			var mobile = self.TraitOrDefault<Mobile>();
			if (mobile != null)
				crushes = mobile.Info.LocomotorInfo.Crushes;
```
```

Reply:
should be all fixed